common/dpcpp/slab_hash.hpp

```diff
@@ -88,16 +158,42 @@ template <typename K, typename T, typename Hash> class SlabHashTable {
 
       sycl::group_barrier(_gr);
     }
-
     return _ans;
   }
 
 private:
+  void alloc_node(sycl::device_ptr<SlabNode<std::pair<K, T>>> &src) {
+    lock();
+    if (src == nullptr) {
+      auto tmp = _heap.malloc_node();
+      *tmp = SlabNode<std::pair<K, T>>({_empty, T()}); //!!!!!!
+      src = tmp;
+    }
+    unlock();
+    _iter = src;
+  }
+
+  void lock() {
+    auto tmp = _hasher(_key);
+    int i = 0;
+    while (
+        sycl::atomic<uint32_t, sycl::access::address_space::global_device_space>((_lock + (tmp / (CHAR_BIT * sizeof(uint32_t)))))
+            .fetch_or(1 << (tmp % (CHAR_BIT * sizeof(uint32_t)))) &
+        (1 << (tmp % (CHAR_BIT * sizeof(uint32_t))))) {
+    }
+  }
+
+  void unlock() {
+    auto tmp = _hasher(_key);
+    sycl::atomic<uint32_t, sycl::access::address_space::global_device_space>((_lock + (tmp / (CHAR_BIT * sizeof(uint32_t)))))
+        .fetch_and(~(1 << (tmp % (CHAR_BIT * sizeof(uint32_t)))));
```

Comment:
Looks like you'd want to have a `CHAR_BIT * sizeof(uint32_t)` constant available